### PR TITLE
Feature/tidy up legacy bits 2845223

### PR DESCRIPTION
**Pull request by yanniboi**

### CODE REVIEW

File: contacts.install [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+  // Test admin theme is enabled.
+  if ($phase == 'install') {
+    /* @var $themeHandler \Drupal\Core\Extension\ThemeHandler */
+    $themeHandler = \Drupal::service('theme_handler');
+
+    if (!$themeHandler->themeExists('contacts_theme')) {
+      // Try to install theme.
+      $success = \Drupal::service('theme_installer')->install(['contacts_theme']);
+
```
Comment (yanniboi) [inline contacts.install:25]: This might be overkill now... What do you think?

File: tests/src/Functional/ContactsDashboardTest.php [hunk 2]
```diff
@@ -11,7 +11,7 @@
  */
 class ContactsDashboardTest extends BrowserTestBase {
 
-  public static $modules = ['user', 'contacts', 'views'];
+  public static $modules = ['user', 'views', 'help', 'block'];
```
Comment (yanniboi) [inline tests/src/Functional/ContactsDashboardTest.php:18]: Help and block are dependencies of the theme.

File: travis-ci/after/environments/drupal-8.sh [hunk 3]
```diff
@@ -40,3 +40,11 @@ function drupal_ti_ensure_module_linked() {
...
 	git apply -v $DRUPAL_TI_DRUPAL_DIR/modules/contacts/travis-ci/merging_data_types-2693081-15_0.patch
 }
+
+#
+# Ensures that the module is linked but not installed.
+#
+function drupal_ti_ensure_module() {
+	# Ensure the module is linked into the code base.
+	drupal_ti_ensure_module_linked
+}
```
Comment (yanniboi) [inline travis-ci/after/environments/drupal-8.sh:50]: This is stopping the module being enabled for tests.

**Review by yanniboi — COMMENTED** [review]
Tests are passing @andrewbelcher what do you think?

File: contacts.install (hunk 1 above)
Comment (andrewbelcher) [reply contacts.install:25]: I think we probably shouldn't do it as part of hook requirements... We could check that it is available for install, but do the actual install as part of `hook_install()`?

File: contacts.install [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      if (!$success) {
+        $requirements['theme'] = array(
```
Comment (andrewbelcher) [inline contacts.install]: `'theme'` is probably too non-specific - `$requirements` gets merged into an array with any other `hook_requirements()` returns, so should probably use `contacts_theme` or something?

File: contacts.install [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+          'title' => t('Missing Theme'),
```
Comment (andrewbelcher) [inline contacts.install]: Can we be slightly more specific in the title? Something like 'Contacts theme missing'?

File: tests/src/Functional/ContactsDashboardTest.php (hunk 2 above)
Comment (andrewbelcher) [reply tests/src/Functional/ContactsDashboardTest.php:18]: This is missing documentation. An `@inheritdoc` with an additional comment as to why we're enabling those modules would be good.

On those requirements, are they hard requirements, or can we put them in as soft dependencies (i.e. wrap their use in checks)?

File: tests/src/Functional/ContactsDashboardTest.php [hunk 6]
```diff
@@ -32,6 +32,7 @@ protected function setUp() {
     $this->adminUser->save();
 
     \Drupal::service('theme_installer')->install(['contacts_theme']);
```
Comment (andrewbelcher) [inline tests/src/Functional/ContactsDashboardTest.php]: With it doing the install as part of `hook_requirements()` (or soon to be `hook_install()`), do we need to do it separately here? And, if not, can `contacts` just go in `self::$modules`?

File: contacts.install (hunk 4 above)
Comment (andrewbelcher) [reply contacts.install]: Need to check that it's not too late as part of `hook_install()` - not sure when the config dependencies are worked out...

File: travis-ci/after/environments/drupal-8.sh (hunk 3 above)
Comment (andrewbelcher) [reply travis-ci/after/environments/drupal-8.sh:50]: Nice :)

**Review by andrewbelcher — CHANGES_REQUESTED** [review]
Couple little changes needed and a few questions I think need answering...

File: contacts.install (hunk 1 above)
Comment (yanniboi) [reply contacts.install:25]: I wonder if hook_install is early enough but can look into that.

File: tests/src/Functional/ContactsDashboardTest.php (hunk 2 above)
Comment (yanniboi) [reply tests/src/Functional/ContactsDashboardTest.php:18]: Config. We add a help block to a region as part of the theme install. We might be able move the block to 'optional'?

Comment (andrewbelcher) [reply tests/src/Functional/ContactsDashboardTest.php:18]: Yes, optional is the answer and then we can ditch them as dependencies at all!

**Comment by yanniboi** [comment]
@andrewbelcher made changes and passing tests. Feel free to review/merge.

File: contacts.install [hunk 7]
```diff
@@ -17,18 +17,28 @@ function contacts_requirements($phase) {
...
-      $success = \Drupal::service('theme_installer')->install(['contacts_theme']);
-
-      if (!$success) {
-        $requirements['theme'] = array(
-          'title' => t('Missing Theme'),
-          'value' => 'To enable the contacts module you must first enable the contacts theme.',
-          'severity' => REQUIREMENT_ERROR,
-        );
-      }
+      $requirements['contacts_theme_missing'] = array(
```
Comment (andrewbelcher) [inline contacts.install]: Hmm... This is going to result in a requirements error if the theme is not installed, which means you now have to install it first. Can we check whether is is available to install before we mark an error?

File: contacts.install [hunk 8]
```diff
@@ -17,18 +17,28 @@ function contacts_requirements($phase) {
...
 
   return $requirements;
 }
+
+/**
+ * Implements hook_module_preinstall().
+ *
+ * Module config depends on a layout provided by the theme.
+ */
+function contacts_module_preinstall() {
```
Comment (andrewbelcher) [inline contacts.install]: [`hook_module_preinstall($module)`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21module.api.php/function/hook_module_preinstall/8.2.x) is called on all installed modules whenever a module is installed, so we need to:

1) Check that this get's called (I'm pretty sure it is as it's added to the module list before the hook is invoked).
2) Add the argument and check that is is `contacts` being installed.
3) Does our test provide coverage for this process?

File: contacts.install (hunk 7 above)
Comment (andrewbelcher) [reply contacts.install]: Think you need to use $theme_data = \Drupal::service('theme_handler')->rebuildThemeData() to see if a theme is available (see \Drupal\Core\Extension\ThemeInstaller::install).

**Review by andrewbelcher — CHANGES_REQUESTED** [review]
Really good step forward, think we're very close!

**Comment by andrewbelcher** [comment]
https://github.com/yanniboi/contacts/pull/11 resolves the remaining changes and adds a bit more test coverage

**Pull request merged**